lib/postprocessing/image.js

```diff
@@ -459,6 +469,35 @@ async function imagePostProcess(intermediateRendition, rendition, directories) {
     await write(img, rendition.path);
 }
 
+async function convertSvg(path, instructions) {
+
+    // IM automatically keeps aspect ratio
+    const width = instructions.width || 400;
+    const height = instructions.height || 400;
+    const fmt = instructions.fmt || "png";
+
+    // Creates intermediate image file since IM operation order matters
+    // Note RSVG is the converter here currently 2020/10/22 causing differences between normal Imagemagic use
+    const image = await read(path);
+
+    if (fmt === "jpg") {
+        image.background("white"); // Change tranparency to white
+    } else {
+        image.background("none"); // Keep alpha background
+    }
+    image.setFormat(fmt);
+    image.rawSize(width, height); // Size of SVG before convertion
+
+    // Future - allow an instruction to define what color should be converted to transparent
+    // Future - handle fonts - currently renders generic blocks for each letter
+    // Future - inkscape - this may cover all filters currently not handled
+
+    path = path.replace(/svg*/, fmt);
+    await write(image, path);
```

Comment:
This runs a _separate_ imagemagick convert command, but we should be able to run everything in the "main" imagemagick command of post processing, by just adding the necessary extra instructions (background and rawSize). This will be faster.

Reply:
Even though some operation need to be applied in order?

Reply:
Imagemagick can handle pretty complex "pipelines" itself in a single command. We just have to set the commands (using `gm.xxxx()`) in the right order. We had the same with the watermarking, this was much more complex and was doable in 1 go using imagemagick's flexible `draw` command. But in this case it's trivial I think, we only have to set the background and rawSize stuff at the beginning (I think).

Probably right at the start of [`applyOperations()`](https://github.com/adobe/asset-compute-sdk/blob/master/lib/postprocessing/image.js#L319):

```js
// list of output formats not supporting alpha channels (transparency)
const NON_ALPHA_FMT = ["jpg"];

async function applyOperations(img, intermediateRendition, rendition, directories) {
    const instructions = rendition.instructions;

    if (intermediateRendition._intermediateMetadata.format === "SVG") {
        if (NON_ALPHA_FMT.includes(fmt)) {
            // Change transparency to white
            image.background("srgb(255,255,255)");
        } else {
            // Keep transparency
            image.background("none");
        }
        // some svgs have no size (only percentage width/height), so we scale them to our target size
        image.rawSize(width, height);
    }

    // handle exif orientation
    // MUST happen before img.resize() for orientation 5 to 8 (otherwise width & height must be swapped)
    // not using img.autoOrient() because that does -strip which strips the entire exif metadata
    // but we only want to remove the exif Orientation flag which we simply unset
    img.out("-auto-orient");
    img.out("-orient");
    img.out("Undefined");

    // ...
```

I also made some more changes
- constant array `NON_ALPHA_FMT` including "jpg" and maybe more in the future
- note the use of `srgb(255,255,255)` instead of `white`. From the [colors.xml discussion](https://git.corp.adobe.com/nui/openwhisk-runtime-nodejs/pull/40#discussion_r3244843).
- some more details in the comments

Reply:
FYI, updated the above comment with `rgba(0, 0, 0, 0.0)` instead of `none`. From [here](https://git.corp.adobe.com/nui/openwhisk-runtime-nodejs/pull/40#issuecomment-2755002).

Reply:
> bit depth going from 8 to 16 ... The files are 2x as large

Would be good to understand why that is happening. I am sure this can be tweaked. I think we should avoid 2x as large.

Is this happening only for SVG input? Or does this happen for all files now? (afaics it shouldn't if we only do those extra settings if input == SVG).

> The bad news is in the scenario of JPG, the background is not forced to white (has nothing to do with the background value being named or rgba values). So I'm guessing it's part of the other ticket you mentioned last week.

What is the background result if not white? In my local testing it worked with white (using `convert -background 'rgb(255,255,255)' input.svg output.jpg`, but that was using the imagemagick on my Mac (with rsvg). So it seems possible, maybe we just have to tweak something in our imagemagick installation on the docker image from https://git.corp.adobe.com/nui/openwhisk-runtime-nodejs/pull/40.

What ticket are you referring to? (Sorry my memory doesn't find a match, but that could be me :-)

Reply:
The resulting image is all black (where the SVG image is only one color of black with transparent background)
Same result as this par of one of the SVG threads https://cq-dev.slack.com/archives/GAMJLHSBH/p1603387248125700?thread_ts=1603384448.078800&cid=GAMJLHSBH

For the ticket, I just realized I didn't create it yet.... but is related to https://cq-dev.slack.com/archives/GAMJLHSBH/p1603387788138800?thread_ts=1603384448.078800&cid=GAMJLHSBH

It looks like none of the renditions currently in the worker tests are 16 bit nor go from 8 to 16 bit. I will look some more to see where maybe the 16 bit get's introduced. Do we want to have a future feature for bit designation in the instructions object?

Reply:
> The resulting image is all black

Would you mind sharing the SVG source image and the actual `convert` command run (if done in sdk/through test-worker, the imagemagick convert command is in the logs)? Maybe best in the jira issue or slack (not sure if the image can be shared publicly here).

> Do we want to have a future feature for bit designation in the instructions object?

Yes, possibly. But here we should keep it simple.

Reply:
So far it seems for the bit depth change:

- Writing the intermediate rendition asset does NOT convert the B&W 8 bit to 16 bit whether or not the `convertSVG()` is called in `applyOperations()` or `imagePostProcess()`
- Not writing the intermediate rendition asset DOES convert to B&W 16 bit whether or not the `convertSVG()` is called in `applyOperations()` or `imagePostProcess()`
- Defining background using `rgbs`/`rgb` vs color name does not make a difference
- This issue is not seen no all B&W SVGs

Here are the Imagemagick commands recorded in the test log maybe some sees something?

New code
```
convert -background 'rgba(0, 0, 0, 0.0)' /nodejsAction/gR2hVJjl/work/kgty20sk/out/rendition0.png[0] -size 1000x400 -auto-orient -orient Undefined -resize 1000> -define png:exclude-chunks=date -colorspace sRGB /nodejsAction/gR2hVJjl/work/kgty20sk/out/post/rendition0.png
```


Old code
```
convert -background 'rgba(0, 0, 0, 0.0)' /nodejsAction/s0BFLo15/work/kgty91pt/out/rendition0.png[0] -size 1000x400 png:-
convert /nodejsAction/s0BFLo15/work/kgty91pt/out/rendition0.png[0] -auto-orient -orient Undefined -resize 1000> -define png:exclude-chunks=date -colorspace sRGB /nodejsAction/s0BFLo15/work/kgty91pt/out/post/rendition0.png
```

Reply:
If I run this locally, I get the same results. The intermediate and the results are always 16 bit PNG, regardless of 1 or 2 step process.

If I have an 8-bit PNG as source (instead of SVG), I get an 8-bit output.

I believe this is simply due to the extra alpha channel we need to preserve the transparency.

Here is what I ran, using `lg-information.svg` from the CQ jira issue:

1. SVG one step
```
> convert -background 'rgba(0, 0, 0, 0.0)' lg-information.svg -size 1000x400 -auto-orient -orient Undefined -resize '1000>' -define png:exclude-chunks=date -colorspace sRGB out.png

> identify out.png 
out.png PNG 216x216 216x216+0+0 16-bit sRGB 13872B 0.000u 0:00.000
```

2. SVG two steps
```
> convert -background 'rgba(0, 0, 0, 0.0)' lg-information.svg -size 1000x400 intermediate.png

> identify intermediate.png 
intermediate.png PNG 216x216 216x216+0+0 16-bit sRGB 13970B 0.010u 0:00.000

> convert intermediate.png -auto-orient -orient Undefined -resize '1000>' -define png:exclude-chunks=date -colorspace sRGB out2.png

> identify out2.png 
out2.png PNG 216x216 216x216+0+0 16-bit sRGB 13872B 0.000u 0:00.000
```

3. PNG 8-bit, one step
```
> identify image.png 
image.png PNG 490x348 490x348+0+0 8-bit sRGB 196756B 0.000u 0:00.000

> convert image.png -auto-orient -orient Undefined -resize '1000>' -define png:exclude-chunks=date -colorspace sRGB out3.png

> identify out3.png 
out3.png PNG 490x348 490x348+0+0 8-bit sRGB 157082B 0.010u 0:00.000
```

Reply:
I did some more testing and the 8-bit instead of 16-bit output happens when _all_ of these are true:
1. if the SVG is pure black & white 
2. `rsvg` is used
3. no `-resize` command is present (!)
4. `-size` comes after the input file in the argument list

Regarding 2: Turns out the imagemagick on my Mac does not use `rsvg`. It's visible using `convert -list configure` (which has `--with-rsvg=no` on my mac), and `convert -list delegate` is misleading).

It seems rsvg is smart and avoids use of alpha channel or extra color bits if not needed due to the SVG being limited to black & white.

Regarding 3: I started with the full command and removed all options to identify that `-resize` makes the difference compared to the 2 step process. I think we could theoretically just have `-size` (in case of a SVG), but we would have to handle the aspect ratio logic ourselves and read the svg size first, which would not always work.

Regarding 4: This actually is not quite optimal as it means it renders the SVG onto a larger canvas first (I think?). At least if you have just `-size` and no `-resize`, the output image will have a different size depending on whether `-size` comes before the input file argument or after. To add, I am not sure if it's controllable with `gm()` which API-wise always starts with the input file.

**tldr;**

Having 16-bit output PNGs from SVG is not an issue. As long as existing conversions from PNG/JPEG etc. aren't affected, which they should not.

But we should check the `-size` position and see if we can get it right - assuming right = in front.

Reply:
Here is what I have done locally
2. I updated the docker image to install RSVG and it's dependents after new image layer is defined (checked into [PR#40](https://git.corp.adobe.com/nui/openwhisk-runtime-nodejs/pull/40)
3. `-resize` is applied outside the scope of SVG with - should I force remove if SVG? or leave as is and live with it as is
```
    if (instructions.width || instructions.height) {
        // set ">" to prevent upscaling
        // docs: https://www.imagemagick.org/script/command-line-processing.php#geometry
        img.resize(instructions.width || null, instructions.height || null, ">");
    }
```
4. Using `gm.in()` to define `-size` forces it to before the source is defined
```
convert -background 'rgba(0, 0, 0, 0.0)' -size 1000x400 /nodejsAction/7p3FcXVQ/work/kgv851kp/out/rendition0.png[0] -colorspace sRGB -auto-orient -orient Undefined -resize 1000> -define png:exclude-chunks=date -colorspace sRGB /nodejsAction/7p3FcXVQ/work/kgv851kp/out/post/rendition0.png
```

Reply:
3. not sure, we should test it well. I would say it's not bad if we leave it in and get a few more 16-bit PNGs for some SVGs, hard to imagine someone would care.

4. cool 👍